File: src/Functional/Elixir/chain_of_responsibility.exs
defmodule HandlerChain do
  def route(amount, handlers) do
    result =
      Enum.reduce_while(handlers, [], fn handler, visited ->
        visited_now = visited ++ [handler.name]

        if handler.accepts.(amount) do
          {:halt, {visited_now, handler.name}}
        else
          {:cont, visited_now}
        end
      end)

    case result do
      {visited, handled} -> {visited, handled}
      _ -> raise "No handler accepted the request"
    end
  end
end

handlers = [
  %{name: "faq", accepts: fn amount -> amount <= 50 end},
  %{name: "billing", accepts: fn amount -> amount <= 500 end},
  %{name: "escalation", accepts: fn _ -> true end}
]

amount = 250
{visited, handled} = HandlerChain.route(amount, handlers)
IO.puts("visited=#{Enum.join(visited, ">");handled=#{handled};result=refund(#{amount})")

File: src/Functional/Erlang/chain_of_responsibility.erl
-module(chain_of_responsibility).
-export([main/0]).

accepts(_Amount, all) -> true;
accepts(Amount, Limit) -> Amount =< Limit.

route(Amount, [{Name, Limit} | Rest], Visited) ->
    VisitedNow = Visited ++ [Name],
    case accepts(Amount, Limit) of
        true -> {VisitedNow, Name};
        false -> route(Amount, Rest, VisitedNow)
    end;
route(_Amount, [], _Visited) ->
    error(no_handler_accepted_request).

main() ->
    Amount = 250,
    Handlers = [{"faq", 50}, {"billing", 500}, {"escalation", all}],
    {Visited, Handled} = route(Amount, Handlers, []),
    io:format("visited=~s;handled=~s;result=refund(~B)~n",
              [string:join(Visited, ">"), Handled, Amount]).
